core/dbt/graph/selector.py

```diff
@@ -162,12 +187,53 @@ def filter_selection(self, selected: Set[UniqueId]) -> Set[UniqueId]:
             unique_id for unique_id in selected if self._is_match(unique_id)
         }
 
-    def expand_selection(self, selected: Set[UniqueId]) -> Set[UniqueId]:
-        """Perform selector-specific expansion."""
+    def expand_selection(
+        self, selected: Set[UniqueId], greedy: bool = False
+    ) -> Tuple[Set[UniqueId], Set[UniqueId]]:
+        # Test selection can expand to include an implicitly/indirectly selected test.
+        # In this way, `dbt test -m model_a` also includes tests that directly depend on `model_a`.
+        # Expansion has two modes, GREEDY and NOT GREEDY.
+        #
+        # GREEDY mode: If ANY parent is selected, select the test. We use this for EXCLUSION.
+        #
+        # NOT GREEDY mode:
+        #  - If ALL parents are selected, select the test.
+        #  - If ANY parent is missing, return it separately. We'll keep it around
+        #    for later and see if its other parents show up.
+        # We use this for INCLUSION.
+
+        direct_nodes = selected
+        indirect_nodes = set()
+
+        for unique_id in self.graph.select_successors(selected):
+            if unique_id in self.manifest.nodes:
+                node = self.manifest.nodes[unique_id]
+                if node.resource_type == NodeType.Test:
+                    # should we add it in directly?
+                    if greedy or set(node.depends_on.nodes) <= set(selected):
+                        direct_nodes.add(unique_id)
+                    # if not:
+                    else:
+                        indirect_nodes.add(unique_id)
+
+        return direct_nodes, indirect_nodes
+
+    def incorporate_indirect_nodes(
+        self, selected: Set[UniqueId], greedy_nodes: Set[UniqueId] = set()
+    ) -> Set[UniqueId]:
+        # Check tests previously selected indirectly to see if ALL their
+        # parents are now present.
+
+        for unique_id in greedy_nodes:
```

Comment:
similar feedback to the side-effect comment above - can we make a copy of `selected` before mutating it?

Reply:
Also, there's some complexity here! We are both adding elements to the `selected` set while also checking if `node.depends_on.nodes` is a subset of `selected`. Is that going to have any funny/unexpected impacts?

I need to dig in here to understand if this would be an issue, but I'm just imagining:

```
selected: {A}

Node A:
 - Depends on: []

Node B:
 - Depends on: [A]

Node C:
 - Depends on [A, B]
```

If `greedy_nodes == [B, C]`, then:
 - loop iter 0: B gets added to `selected`
 - loop iter 1: C gets added to `selected`
 - Return value is `{A, B, C}`

If the order is reversed and `greedy_nodes == [C, B]`, then:
 - loop iter 0: C does not get added to `selected`
 - loop iter 1: B does not get added to `selected`
 - Return value is `{A}`

I actually... don't know if that's an issue based on how we use this method, would need to dig in a lot further to understand, but wanted to mention it in case it sparks any thoughts!

Reply:
Definitely not my intention here for order to matter here. I rewrote so that it instantiates and mutates a different variable from the value passed in as the argument. A test cannot be parent to another test, so for the specific case tackled in this PR, I think we're in the clear—but this was a great thing to call out and change, in the spirit of making indirect selection as much of its own thing as possible, not uniquely suited to tests.